### DISCUSSION TITLE
PageRank Alogrithm

### DISCUSSION THREAD

**Discussion by jedi1992** [EvgSkv/logica#106] [Q&A] [answered]
Hi 
I was going through your example here https://twitter.com/EvgSkvDev/status/1396168603217448962

a) I guess Vertex has as many entries for each node as it participates  in a link - 1 for source and 1 for sink  
b) Degree is the number of times  a node is a   source
c) PageRank calculated with first Rule is just counting number of times a node occurs in Vertex. So how many links it is a  part of. 
d) PageRank calculated second time is about getting some  reflected glory  from the Source , added upto its previous points. 

I added some links where in there is some recursion - ie Node 1 is source of Node 8  and again Node  8 is source of Node 1.  So how does the query know not to get caught in the loop.  Or should I mentally look at it as - take the node in question  , take a snapshot of the DB and then update only this particular node. Again repeat for other nodes. 
Does that mean if it is doing in some order of execution do the earlier nodes miss the benefit ? 
i.e either Node1 or Node8 will not get the latest points if there is an order for execution.

**Comment by EvgSkv** [comment, marked as answer]
Hi,
your understanding is going in the right direction!
It's just computation is done iteratively for all of the nodes.
I.e. at first each node gets rank of one from the rule:
```
PageRank(a) += 1.0 :- Vertex(a);                                                                                                                    
```
It literally gets compiled as
```
PageRank_r0(a) += 1.0 :- Vertex(a);                                                                                                                    
```

Then first iteration adds the rank of one what you get from r0:
```
PageRank_r1(a) += 1.0 :- Vertex(a);                                                                                                                    
PageRank_r1(a) += 0.5 * PageRank_r0(b) / Degree(b) :-                                                                                                     
   Link(b, a);    
```
And so on.
It's up to the user to set the recursion depth with `@Recursive` annotation to achieve reasonable conversion rate. Default depth is 8, which seems to be good enough for PageRank.
See [CoLab](https://colab.research.google.com/github/EvgSkv/logica/blob/main/examples/more/Page_Rank.ipynb) with the example.

I've also created [a tweet](https://twitter.com/EvgSkvDev/status/1401294493274443777) about it :-)

Let me know if you have further questions.

**Reply by jedi1992** [reply]
Yes very clear.  Thanks a ton 👍
